syntax = "proto3";

package rat;

message Packet {
  enum Type {
    UNKNOWN = 0;
    LIST_FILES = 1;
    READ_FILE = 2;
    TRANSFER_FILE = 3;
    LIST_PROCESSES = 4;
    KILL_PROCESS = 5;
    IDENTIFY = 6;
    ACK = 7;
    ERROR = 8;
  }
  Type type = 1;
  string packet_id = 2;
  string source_id = 3;
  string destination_id = 4;
  bool encrypted = 5;
  oneof payload {
    FileChunk file_chunk = 6;
    CommandData command_data = 7;
    ResponseData response_data = 8;
  }
}

message FileChunk {
  string file_id = 1;
  uint64 sequence_number = 2;
  uint64 total_packets = 3;
  string file_name = 4;
  bytes data = 5;
}

message CommandData {
  string command = 1;
}

message ResponseData {
  bool success = 1;
  bytes data = 2;
  string error_message = 3;
}